### DISCUSSION TITLE
在打印边时，起点和终点都是自动生成的id，如何同时打印自定义起点和终点号呢？

### DISCUSSION THREAD

**Discussion by zjarles** [alibaba/GraphScope#3187] [Q&A] [answered]
执行g.V().has('trmnl','trmnl_black_flag',1).outE('register').dedup().limit(2).to_list()的结果为：
[e[6917529027641081901][-7493989779944505311-register->8651414884178722861],
 e[-8070450532247928817][2305843009213693983-register->-7489486180317134816]]
这个自动生成的id：6917529027641081901，没有业务含义，如何打印点trmnl的终端号呢，点trmnl的载入csv第1列就是'终端号'，如何同时显示起点的终端号，终点的注册号呢？

oid_type: STRING
vid_type: ULONG
type: VERTEX
Label: trmnl
Properties: Property(0, trmnl_black_flag, LONG, False, ), Property(1, meid, STRING, False, )
Comment: 

为了解决这个问题，在graph = session.g(oid_type="string")中，加上了generate_eid=False，但上述执行语句报错，无法执行，所以没有解决上述问题。

**Comment by sighingnow** [comment]
GIE doesn't work `generate_eid=False`.

> 如何同时显示起点的终端号，终点的注册号呢？

Try `.valueMap()`.

**Comment by zjarles** [comment]
您好，
&nbsp; &nbsp; &nbsp; &nbsp;加上.valueMap().后是，print(g.V().has('trmnl','trmnl_black_flag',1).outE('register').dedup().limit(10).valueMap().to_list())
执行结果如下：
[{'eid': [1152921504606847060]}, {'eid': [576460752303423561]}, {'eid': [1729382256910270544]}, {'eid': [7493989779944505386]}, {'eid': [5188146770730811456]}, {'eid': [6917529027641081896]}, {'eid': [64]}, {'eid': [1729382256910270558]}, {'eid': [1152921504606847011]}, {'eid': [2305843009213693997]}]



还是没有打印终端号，结果都是自动生成的id。请问有其他办法吗？




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "alibaba/GraphScope"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年9月6日(星期三) 下午3:23
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [alibaba/GraphScope] 在打印边时，起点和终点都是自动生成的id，如何同时打印自定义起点和终点号呢？ (Discussion #3187)





 
GIE doesn't work generate_eid=False.
  
如何同时显示起点的终端号，终点的注册号呢？
  
Try .valueMap().

**Reply by zjarles** [reply]
g.V().has('trmnl','trmnl_black_flag',1).both('1..2', 'register').dedup().valueMap().limit(10).to_list()   -- 这样可以显示 终点的注册号，但没有起点，如何同时显示起点和终点的业务号呢？

**Reply by BingqingLyu** [reply, marked as answer]
Hi, for the query:
```
g.V().has('trmnl','trmnl_black_flag',1).both('1..2', 'register')
```
You'll get the end vertex of the path by default. That's why you only get properties of the end vertex when you further query with `valueMap()`.

If you want to get properties of start vertex and end vertex of the path, you can try the following query:
```
g.V().has('trmnl','trmnl_black_flag',1).as('start').both('1..2', 'register').dedup().endV().as('end').select('start','end').by(valueMap()).limit(10)
```

**Comment by zjarles** [comment]
感谢，那如果想打印3-4跳的中间路径，如何既显示起终点，又显示中间的业务号呢？我尝试了&nbsp;
interactive.execute("g.V().has('trmnl','trmnl_black_flag',1).as('start').both('3..4', 'register').with('PATH_OPT', 'ARBITRARY') \
.with('RESULT_OPT', 'ALL_V').as('allV').select('start','allV'))").one()

结果为：
[{'start': v[2305843009213693975],   'allV': [v[2305843009213693975],    v[2886807361144487959],    v[2305843009213693975],    v[2886807361144487959]]},  {'start': v[1152921504606846998],   'allV': [v[1152921504606846998],    v[4503599627370521],    v[1152921504606846998],    v[4503599627370521]]}]不显示业务号，是 自动生成的id ,加上 .by(valueMap())后：
interactive.execute("g.V().has('trmnl','trmnl_black_flag',1).as('start').both('3..4', 'register').with('PATH_OPT', 'ARBITRARY') .with('RESULT_OPT', 'ALL_V').as('allV').select('start','allV').by(valueMap())").one()

结果 'allV' 只有一个业务号，二跳和三跳的业务号消失了，只显示末端终点的业务号。如何把路径上 起始点、二跳点、三跳点、终点的完整路径以“业务号“显示出来呢？


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "alibaba/GraphScope"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年9月6日(星期三) 晚上6:15
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [alibaba/GraphScope] 在打印边时，起点和终点都是自动生成的id，如何同时打印自定义起点和终点号呢？ (Discussion #3187)





 
Hi, for the query:
 g.V().has('trmnl','trmnl_black_flag',1).both('1..2', 'register')  
You'll get the end vertex of the path by default. That's why you only get properties of the end vertex when you further query with valueMap().
 
If you want to get properties of start vertex and end vertex of the path, you can try the following query:
 g.V().has('trmnl','trmnl_black_flag',1).as('start').both('1..2', 'register').dedup().endV().as('end').select('start','end').by(valueMap()).limit(10)

**Reply by BingqingLyu** [reply]
您好，目前我们还暂未支持通过`valueMap()`算子来打印path中每个点的属性这样的能力。可能有两种解决方式：
（1）您可以先找到符合条件的path，得到path结果后，再对其中的点查询其设备号（例如，通过 `g.V(2305843009213693975).valueMap()` 这样的方式查询）；
（2）如果确实您有强需求，能否麻烦提个issue，我们作为一个新功能来排期支持。感谢您的关注！

**Comment by zjarles** [comment]
那在建图的时候，点trmnl的id是否能不用自动生成的id，而是用终端号来当作点的id呢？终端号本身就是唯一的，没必要再自动生成id了。这样打印路径的时候，显示的id，就是终端号了。
我尝试过 令
graph = session.g(oid_type="string",generate_eid=False)&nbsp; # 加上generate_eid=False
node = Loader("./abc.csv",delimiter='\t', header_row=True,key='meid')，加上key='meid'
也没有用处。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "alibaba/GraphScope"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年9月7日(星期四) 中午12:17
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [alibaba/GraphScope] 在打印边时，起点和终点都是自动生成的id，如何同时打印自定义起点和终点号呢？ (Discussion #3187)





 
Hi, we have not supported a step that can project the properties of every vertex on the path yet.
 However, you can try to obtain the properties in another query, i.e., find the path first, and then for each vertex with gid (the automatically generate vertex id), you can query g.V(gid).valueMap() flexibly.

**Reply by BingqingLyu** [reply]
抱歉，我们并不能将终端号的id来作为内部的唯一标识。虽然对于`trmnl`这种类型的点来说，终端号是它的唯一标志符；但是在全图上，例如你有另一种类型的点`trmnl2`，它的`终端号2`可能跟`trmnl`类型点的`终端号`是会重复的。所以，我们没办法用某个类型的唯一标志符，在全图上来唯一标志。
ref: https://graphscope.io/docs/interactive_engine/tinkerpop/faq#whats-the-difference-between-inner-id-and-property-id

**Reply by longbinlai** [reply]
@zjarles https://github.com/alibaba/GraphScope/pull/3213, 我们现在提供了这个功能。稍后，可以安装试用版：pip3 install graphscope --pre ，您可以试试看。
